### DISCUSSION TITLE
Is it possible to include support for CrmServiceClient ?

### DISCUSSION THREAD

**Discussion by pavansarma93** [jordimontana82/fake-xrm-easy#581] [Q&A]
Currently I am having an azure function V1.1 which uses Crmserviceclient to perform Dynamics operations. However all the stub methods available in FakeXrmEasy are plugin or workflow related.

Is it possible to provide support for CrmServiceClient ?

Here is a sample code to test I am using

`using System;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;
using Microsoft.Azure.WebJobs.Host;
using Microsoft.Xrm.Sdk.Query;
using Microsoft.Xrm.Tooling.Connector;

namespace FnApp_Framework
{
    public static class Function1
    {
        [FunctionName("Function1")]
        public static async Task<HttpResponseMessage> Run([HttpTrigger(AuthorizationLevel.Function, "get", "post", Route = null)]HttpRequestMessage req, TraceWriter log)
        {
            log.Info("C# HTTP trigger function processed a request.");
            string name = req.GetQueryNameValuePairs()
                .FirstOrDefault(q => string.Compare(q.Key, "name", true) == 0)
                .Value;
     try
            {
                string con = @"
            AuthType = ClientSecret;
            ClientId = clientid;
            ClientSecret = clientsec;
            Url = https://crm.crm11.dynamics.com/;";

                //System.Net.WebProxy
                CrmServiceClient service = new CrmServiceClient(con);

                if (service.IsReady)
                {
                    QueryExpression query = new QueryExpression("account");
                    service.RetrieveMultiple(query);
                    log.Info("ready");
                }
            }
            catch(Exception ex)
            {
                log.Info(ex.Message);
                log.Info(ex.StackTrace);
            }
                if (name == null)
            {
                // Get request body
                dynamic data = await req.Content.ReadAsAsync<object>();
                name = data?.name;
            }

            return name == null
                ? req.CreateResponse(HttpStatusCode.BadRequest, "Please pass a name on the query string or in the request body")
                : req.CreateResponse(HttpStatusCode.OK, "Hello " + name);
        }
    }
}`

**Comment by jordimontana82** [comment]
HI @pavansarma93 , 

It is already there.

Since FakeXrmEasy provides mocks for the IOrganizationService interface, and CrmServiceClient implements such interface, you could also mock these requests too.

The plugin and workflow helpers are, in fact, helpers on top of that functionality. The most valuable aspect of FakeXrmEasy is it's low-level mocking and in-memory query engine.

Hope this helps!

**Reply by jordimontana82** [reply]
Just refactor your Azure Function code so it relies on the interface only, and pass the CrmServiceClient for production code and the OrganizationService retrieved via GetOrganizationService() in your test project.

**Reply by jordimontana82** [reply]
@pavansarma93  did this help?

**Reply by jordimontana82** [reply]
@pavansarma93  Could you mark this question as answered, please? Also, please check out this announcement: https://github.com/jordimontana82/fake-xrm-easy/issues/582

**Reply by pavansarma93** [reply]
@jordimontana82 : I am new to this so any example would be helpful. Because I have created unit test for plugins with the help of examples from your blog only

**Reply by jordimontana82** [reply]
@pavansarma93 You'll need to: 

   - 1) Refactor the code in the Azure Function to a separate class that only references IOrganizationService interface like
       
 
```csharp
public class MyAzureFunctionLogic {
           public static EntityCollection GetQueryResults(IOrganizationService service) {
                  QueryExpression query = new QueryExpression("account");
                  return service.RetrieveMultiple(query);
                  
           }
       }
```

   - 2) Add a unit test project that references such project, and use the GetOrganizationService method to retrieve a fake IOrganizationService interface, then pass it to your azure function like the sample unit test below:

```csharp
    [Fact]
    public void Should_return_data() {
          var ctx = new XrmFakedContext();
          var service = ctx.GetOrganizationService();
          ctx.Initialize(new List<Entity>() {  

              //some test entities 
             new Account() { Id = Guid.NewGuid() },
             new Account() { Id = Guid.NewGuid() },

          };
          var result = MyAzureFunctionLogic.GetQueryResults(service);
          Assert.Equal (2, result.Entities.Count)  //will depend on data
    }
```

  - 3) The production code in your Azure function will need to pass the CrmServiceClient (like below) at some stage before that function is called, but in the unit test project will use the fake IOrganizationService will be used.

```csharp
    CrmServiceClient service = new CrmServiceClient(con); 
    if (service.IsReady)
    {
        MyAzureFunctionLogic.GetQueryResults(service);
    }
```


Does this make sense?

**Reply by pavansarma93** [reply]
@jordimontana82 : Ok , so what you are saying is we cannot execute this similar to plugin (complete plugin would be ran through ) and only can TEST the code referred separately.

Thank you

**Reply by jordimontana82** [reply]
Kind of, you could put of course the entire azure function code in a single .Execute method, like a plugin, the only requirement is that the code must reference IOrganizationService as opposed to their implementation (CrmServiceClient) which will be passed in at the begining of the function exection like shown above.  Pretty much any logic you can execute with CrmServiceClient can also be executed via IOrganizationService.... Makes sense?

**Reply by jordimontana82** [reply]
@pavansarma93  Does that answer your question?

**Reply by sandeep-chinni** [reply]
@jordimontana82 I have a AzureFunction implementation that does not use OrganizationService at all. I use an implementation of ServiceClient(Microsoft.PowerPlatform.Dataverse.Client.ServiceClient) instead.

Based on what you have mentioned above, unless I rework my implementation to use OrganizationService, I cannot use FakeXrmEasy?
I have been able to write tests using Moq mocks, but I see the value in using FakeXrmEasy because of its in-memory database. I mean it is essentially like a Moq, but maybe a little more cleaner?

**Reply by jordimontana82** [reply]
Hi Sandeep @sandeep-chinni  ,  

There is a newer version of the framework that supports .net core and the new Dataverse client already implemented for you here: https://github.com/DynamicsValue/fake-xrm-easy. 

Here's a new docs site to get you started: https://dynamicsvalue.github.io/fake-xrm-easy-docs/quickstart/


**Important**: That new implementation is still open source and free of charge for open source repositories and non-commercial use, but requires a small fee for commercial purposes in private repos: https://dynamicsvalue.github.io/fake-xrm-easy-docs/licensing/

**Why is important**: The commercial license supports a sustainable OSS model and is going to significantly be more cost effective for your company than yourself: 

- Upgrading and **porting FakeXrmEasy to .net core** and the new DataverseClient
- **Implement and maintain parallel versions of FakeXrmEasy** for both server side development (because DataverseClient is not meant to be used for that), and client-side development with DataverseClient, until, eventually, there is full parity (**in the fullness of time...**)
- All the new features, fixes and security patches we'll be adding that will be covered in your subscription and that you, again, won't have to implement
- Up to date documentation and release notes
- Basic support and answering support feedback, and questions (like this one)

Most of the api in v1.x is forward compatible to v2.x and v3.x, so the transition should be fairly straightforward.

**Reply by sandeep-chinni** [reply]
@jordimontana82 Thank you for the response. I have started a discussion in the repo for the newer version of the framework. Here: https://github.com/DynamicsValue/fake-xrm-easy/discussions/29

Using the newer repo has not resolved my issue or at least I don't know how.

**Reply by jordimontana82** [reply]
Thanks, will carry on in the thread, cheers

**Reply by KannanPooja** [reply]
Any update on this?
I am not able to set IOrganisationservice, to this CrmServiceClient.
. Net core Azure Function  code.
Thank you very much.

**Reply by jordimontana82** [reply]
Hi @KannanPooja  this repo is deprecated, it won't receive further improvements like that one. This discussion has been moved to the new repo (with a different licensing model) here: https://github.com/DynamicsValue/fake-xrm-easy/discussions/29 

If you check that other discussion, you'll the new version already implements ServiceClient, most of the time you'll just need the interfaces, we're discussing adding some initial support for extension methods that only live in the ServiceClient class and that are not defined at the interface level. For extension methods like these it would make sense, otherwise the interfaces in the new version of FakeXrmEasy v3.x should work fine.

If you're already a commercial customer, then please drop us an email at https://dynamicsvalue.com/contact.

Thanks!
